src/tests/decider-distro-tests-rs/tests/test.rs

```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
+    }
+
+    server.shutdown().await;
+}
+
+/// Test worker registering scenarios  
+///
+/// Note that atm *Decider* doesn't process the case when worker registration fails
+/// the deal is joined nevertheless
+///
+#[tokio::test]
+async fn test_register_worker_fails() {
+    enable_decider_logs();
+    const BLOCK_INIT: u32 = 1;
+    const DEAL_ID: &'static str = DEAL_IDS[0];
+    const BLOCK_NUMBER: u32 = 32;
+    const BLOCK_NUMBER_LATER: u32 = 200;
+
+    let (mut server, mut recv_request, send_response) = run_test_server();
+    let url = server.url.clone();
+
+    let empty_config = TriggerConfig::default();
+    let mut distro = make_distro_with_api_and_config(url, empty_config);
+    let swarms = make_swarms_with_cfg(1, move |mut cfg| {
+        cfg.enabled_system_services = vec!["aqua-ipfs".to_string()];
+        cfg.extend_system_services = vec![distro.clone()];
+        cfg
+    })
+    .await;
+    let mut client = ConnectedClient::connect_with_keypair(
+        swarms[0].multiaddr.clone(),
+        Some(swarms[0].management_keypair.clone()),
+    )
+    .await
+    .unwrap();
+
+    update_decider_script_for_tests(&mut client, swarms[0].tmp_dir.clone()).await;
+    // Initial run for installing the first deal
+    update_config(&mut client, &oneshot_config()).await.unwrap();
+    {
+        let error_value = json!({
+            "code": -32000,
+            "message": "intentional error",
+        });
+        // Reqs: blockNumber, getLogs and one of gasPrice, getTransactionCount and sendRawTransaction
+        // (try all to not depend on the order)
+        for step in 0..3 {
+            let (method, params) = recv_request.recv().await.unwrap();
+            let response = match method.as_str() {
+                "eth_blockNumber" => Ok(json!(to_hex(BLOCK_INIT))),
+                "eth_getLogs" => {
+                    let log = serde_json::from_value::<LogsReq>(params[0].clone()).unwrap();
+                    Ok(json!([TestApp::log_test_app2(
+                        DEAL_ID,
+                        BLOCK_NUMBER,
+                        log.topics[1].as_str()
+                    )]))
+                }
+                "eth_sendRawTransaction" => Err(error_value.clone()),
+                "eth_getTransactionCount" => Err(error_value.clone()),
+                "eth_gasPrice" => Err(error_value.clone()),
+                _ => panic!("unexpected method: {}", method),
+            };
+            send_response.send(response).unwrap();
+        }
+    }
+    wait_decider(&mut client).await;
+    let deals = get_joined_deals(&mut client).await;
+    assert!(
+        deals.is_empty(),
+        "since the registration failed, the deal must not be joined"
+    );
+
+    update_config(&mut client, &oneshot_config()).await.unwrap();
+    {
+        // Reqs: blockNumber, getLogs, gasPrice, getTransactionCount and sendRawTransaction
+        for _ in 0..5 {
+            let (method, params) = recv_request.recv().await.unwrap();
```

Comment:
What if decider breaks, will it hang forever?

Does decider ever breaks in these tests, ever?